File: library/src/main/java/com/cunoraz/gifview/library/GifView.java
package com.cunoraz.gifview.library;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Movie;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by Cuneyt on 4.10.2015.
 * Gifview
 */
public class GifView extends View {

    private static final int DEFAULT_MOVIE_VIEW_DURATION = 1000;

    private int mMovieResourceId;
    private Movie movie;

    private long mMovieStart;
    private int mCurrentAnimationTime;


    /**
     * Position for drawing animation frames in the center of the view.
     */
    private float mLeft;
    private float mTop;

    /**
     * Scaling factor to fit the animation within view bounds.
     */
    private float mScale;

    /**
     * Scaled movie frames width and height.
     */
    private int mMeasuredMovieWidth;
    private int mMeasuredMovieHeight;

    private volatile boolean mPaused;
    private boolean mVisible = true;

    public GifView(Context context) {
        this(context, null);
    }

    public GifView(Context context, AttributeSet attrs) {
        this(context, attrs, R.styleable.CustomTheme_gifViewStyle);
    }

    public GifView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        setViewAttributes(context, attrs, defStyle);
    }

    @SuppressLint("NewApi")
    private void setViewAttributes(Context context, AttributeSet attrs, int defStyle) {

        /**
         * Starting from HONEYCOMB(Api Level:11) have to turn off HW acceleration to draw
         * Movie on Canvas.
         */
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        }

        final TypedArray array = context.obtainStyledAttributes(attrs,
                R.styleable.GifView, defStyle, R.style.Widget_GifView);

        //-1 is default value
        mMovieResourceId = array.getResourceId(R.styleable.GifView_gif, -1);
        mPaused = array.getBoolean(R.styleable.GifView_paused, false);

        array.recycle();

        if (mMovieResourceId != -1) {
            movie = Movie.decodeStream(getResources().openRawResource(mMovieResourceId));
        }
    }

    public void setGifResource(int movieResourceId) {
        this.mMovieResourceId = movieResourceId;
        movie = Movie.decodeStream(getResources().openRawResource(mMovieResourceId));
        requestLayout();
    }

    public int getGifResource() {

        return this.mMovieResourceId;
    }


    public void play() {
        if (this.mPaused) {
            this.mPaused = false;

            /**
             * Calculate new movie start time, so that it resumes from the same
             * frame.
             */
            mMovieStart = android.os.SystemClock.uptimeMillis() - mCurrentAnimationTime;

            invalidate();
        }
    }

    public void pause() {
        if (!this.mPaused) {
            this.mPaused = true;

            invalidate();
        }

    }


    public boolean isPaused() {
        return this.mPaused;
    }

    public boolean isPlaying() {
        return !this.mPaused;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {

        if (movie != null) {
            int movieWidth = movie.width();
            int movieHeight = movie.height();

			/*
             * Calculate horizontal scaling
			 */
            float scaleH = 1f;
            int measureModeWidth = MeasureSpec.getMode(widthMeasureSpec);

            if (measureModeWidth != MeasureSpec.UNSPECIFIED) {
                int maximumWidth = MeasureSpec.getSize(widthMeasureSpec);
                if (movieWidth > maximumWidth) {
                    scaleH = (float) movieWidth / (float) maximumWidth;
                }
            }

			/*
             * calculate vertical scaling
			 */
            float scaleW = 1f;
            int measureModeHeight = MeasureSpec.getMode(heightMeasureSpec);

            if (measureModeHeight != MeasureSpec.UNSPECIFIED) {
                int maximumHeight = MeasureSpec.getSize(heightMeasureSpec);
                if (movieHeight > maximumHeight) {
                    scaleW = (float) movieHeight / (float) maximumHeight;
                }
            }

			/*
             * calculate overall scale
			 */
            mScale = 1f / Math.max(scaleH, scaleW);

            mMeasuredMovieWidth = (int) (movieWidth * mScale);
            mMeasuredMovieHeight = (int) (movieHeight * mScale);

            setMeasuredDimension(mMeasuredMovieWidth, mMeasuredMovieHeight);

        } else {
			/*
			 * No movie set, just set minimum available size.
			 */
            setMeasuredDimension(getSuggestedMinimumWidth(), getSuggestedMinimumHeight());
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
		/*
		 * Calculate mLeft / mTop for drawing in center
		 */
        mLeft = (getWidth() - mMeasuredMovieWidth) / 2f;
        mTop = (getHeight() - mMeasuredMovieHeight) / 2f;

        mVisible = getVisibility() == View.VISIBLE;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (movie != null) {
            if (!mPaused) {
                updateAnimationTime();
                drawMovieFrame(canvas);
                invalidateView();
            } else {
                drawMovieFrame(canvas);
            }
        }
    }

    /**
     * Invalidates view only if it is mVisible.
     * <br>
     * {@link #postInvalidateOnAnimation()} is used for Jelly Bean and higher.
     */
    @SuppressLint("NewApi")
    private void invalidateView() {
        if (mVisible) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                postInvalidateOnAnimation();
            } else {
                invalidate();
            }
        }
    }

    /**
     * Calculate current animation time
     */
    private void updateAnimationTime() {
        long now = android.os.SystemClock.uptimeMillis();

        if (mMovieStart == 0) {
            mMovieStart = now;
        }

        int dur = movie.duration();

        if (dur == 0) {
            dur = DEFAULT_MOVIE_VIEW_DURATION;
        }

        mCurrentAnimationTime = (int) ((now - mMovieStart) % dur);
    }

    /**
     * Draw current GIF frame
     */
    private void drawMovieFrame(Canvas canvas) {

        movie.setTime(mCurrentAnimationTime);

        canvas.save(Canvas.MATRIX_SAVE_FLAG);
        canvas.scale(mScale, mScale);
        movie.draw(canvas, mLeft / mScale, mTop / mScale);
        canvas.restore();
    }

    @SuppressLint("NewApi")
    @Override
    public void onScreenStateChanged(int screenState) {
        super.onScreenStateChanged(screenState);
        mVisible = screenState == SCREEN_STATE_ON;
        invalidateView();
    }

    @SuppressLint("NewApi")
    @Override
    protected void onVisibilityChanged(View changedView, int visibility) {
        super.onVisibilityChanged(changedView, visibility);
        mVisible = visibility == View.VISIBLE;
        invalidateView();
    }

    @Override
    protected void onWindowVisibilityChanged(int visibility) {
        super.onWindowVisibilityChanged(visibility);
        mVisible = visibility == View.VISIBLE;
        invalidateView();
    }
}

File: library/src/main/java/com/cunoraz/gifview/library/GifView.kt
package com.cunoraz.gifview.library

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Movie
import android.os.Build
import android.util.AttributeSet
import android.view.View

/**
 * Created by Cuneyt on 4.10.2015.
 * Updated by Cuneyt on 02.04.2019.
 * Gifview
 */
class GifView @JvmOverloads
constructor(context: Context,
            attrs: AttributeSet? = null,
            defStyle: Int = R.styleable.CustomTheme_gifViewStyle)
    : View(context, attrs, defStyle) {

    private var movieMovieResourceId: Int = 0
    private var movie: Movie? = null
    private var movieStart: Long = 0
    private var currentAnimationTime: Int = 0

    private var movieLeft: Float = 0F
    private var movieTop: Float = 0F
    private var movieScale: Float = 0F

    private var movieMeasuredMovieWidth: Int = 0
    private var movieMeasuredMovieHeight: Int = 0

    @Volatile
    var isPaused: Boolean = false

    private var isVisible = true

    var gifResource: Int
        get() = this.movieMovieResourceId
        set(movieResourceId) {
            this.movieMovieResourceId = movieResourceId
            movie = Movie.decodeStream(resources.openRawResource(movieMovieResourceId))
            requestLayout()
        }

    val isPlaying: Boolean
        get() = !this.isPaused

    init {
        setViewAttributes(context, attrs, defStyle)
    }

    @SuppressLint("NewApi")
    private fun setViewAttributes(context: Context, attrs: AttributeSet?, defStyle: Int) {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null)

        val array = context.obtainStyledAttributes(attrs,
                R.styleable.GifView, defStyle, R.style.Widget_GifView)

        //-1 is default value
        movieMovieResourceId = array.getResourceId(R.styleable.GifView_gif, -1)
        isPaused = array.getBoolean(R.styleable.GifView_paused, false)

        array.recycle()

        if (movieMovieResourceId != -1) {
            movie = Movie.decodeStream(resources.openRawResource(movieMovieResourceId))
        }
    }


    fun play() {
        if (this.isPaused) {
            this.isPaused = false
            /**
             * Calculate new movie start time, so that it resumes from the same
             * frame.
             */
            movieStart = android.os.SystemClock.uptimeMillis() - currentAnimationTime
            invalidate()
        }
    }

    fun pause() {
        if (!this.isPaused) {
            this.isPaused = true
            invalidate()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (movie != null) {
            val movieWidth = movie!!.width()
            val movieHeight = movie!!.height()
            /*
             * Calculate horizontal scaling
			 */
            var scaleH = 1f
            val measureModeWidth = View.MeasureSpec.getMode(widthMeasureSpec)

            if (measureModeWidth != View.MeasureSpec.UNSPECIFIED) {
                val maximumWidth = View.MeasureSpec.getSize(widthMeasureSpec)
                if (movieWidth > maximumWidth) {
                    scaleH = movieWidth.toFloat() / maximumWidth.toFloat()
                }
            }
            /*
             * calculate vertical scaling
			 */
            var scaleW = 1f
            val measureModeHeight = View.MeasureSpec.getMode(heightMeasureSpec)

            if (measureModeHeight != View.MeasureSpec.UNSPECIFIED) {
                val maximumHeight = View.MeasureSpec.getSize(heightMeasureSpec)
                if (movieHeight > maximumHeight) {
                    scaleW = movieHeight.toFloat() / maximumHeight.toFloat()
                }
            }
            /*
             * calculate overall scale
			 */
            movieScale = 1f / Math.max(scaleH, scaleW)
            movieMeasuredMovieWidth = (movieWidth * movieScale).toInt()
            movieMeasuredMovieHeight = (movieHeight * movieScale).toInt()
            setMeasuredDimension(movieMeasuredMovieWidth, movieMeasuredMovieHeight)
        } else {
            /*
			 * No movie set, just set minimum available size.
			 */
            setMeasuredDimension(suggestedMinimumWidth, suggestedMinimumHeight)
        }
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        /*
         * Calculate movieLeft / movieTop for drawing in center
         */
        movieLeft = (width - movieMeasuredMovieWidth) / 2f
        movieTop = (height - movieMeasuredMovieHeight) / 2f
        isVisible = visibility == View.VISIBLE
    }

    override fun onDraw(canvas: Canvas) {
        if (movie != null) {
            if (!isPaused) {
                updateAnimationTime()
                drawMovieFrame(canvas)
                invalidateView()
            } else {
                drawMovieFrame(canvas)
            }
        }
    }

    /**
     * Invalidates view only if it is isVisible.
     * <br></br>
     * [.postInvalidateOnAnimation] is used for Jelly Bean and higher.
     */
    @SuppressLint("NewApi")
    private fun invalidateView() {
        if (isVisible) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                postInvalidateOnAnimation()
            } else {
                invalidate()
            }
        }
    }

    /**
     * Calculate current animation time
     */
    private fun updateAnimationTime() {
        val now = android.os.SystemClock.uptimeMillis()

        if (movieStart == 0L) {
            movieStart = now
        }

        var duration = movie!!.duration()

        if (duration == 0) {
            duration = DEFAULT_MOVIE_VIEW_DURATION
        }

        currentAnimationTime = ((now - movieStart) % duration).toInt()
    }

    /**
     * Draw current GIF frame
     */
    private fun drawMovieFrame(canvas: Canvas) {
        movie!!.setTime(currentAnimationTime)
        canvas.save()
        canvas.scale(movieScale, movieScale)
        movie!!.draw(canvas, movieLeft / movieScale, movieTop / movieScale)
        canvas.restore()
    }

    @SuppressLint("NewApi")
    override fun onScreenStateChanged(screenState: Int) {
        super.onScreenStateChanged(screenState)
        isVisible = screenState == View.SCREEN_STATE_ON
        invalidateView()
    }

    @SuppressLint("NewApi")
    override fun onVisibilityChanged(changedView: View, visibility: Int) {
        super.onVisibilityChanged(changedView, visibility)
        isVisible = visibility == View.VISIBLE
        invalidateView()
    }

    override fun onWindowVisibilityChanged(visibility: Int) {
        super.onWindowVisibilityChanged(visibility)
        isVisible = visibility == View.VISIBLE
        invalidateView()
    }

    companion object {

        private const val DEFAULT_MOVIE_VIEW_DURATION = 1000
    }
}